src/code/audio_thread_manager.c

```diff
@@ -1,91 +1,150 @@
+ * Original name: audioMgr.c
+ *
+ * This file implements basic thread features for the audio driver. It manages updating the driver on vertical retrace
+ * and sending the audio rsp tasks generated by the driver to the task scheduler.
+ */
+
 #include "global.h"
 
-void func_800C3C80(AudioMgr* audioMgr) {
+void AudioMgr_NotifyTaskDone(AudioMgr* audioMgr) {
     AudioTask* task = audioMgr->rspTask;
 
+    // If the audio rsp task has a message queue to receive task done notifications, post a message to it.
     if (audioMgr->rspTask->msgQueue != NULL) {
         osSendMesg(task->msgQueue, NULL, OS_MESG_BLOCK);
     }
 }
 
+/**
+ * Handle retrace event.
+ * Update the audio driver and schedule audio rsp tasks.
+ */
 void AudioMgr_HandleRetrace(AudioMgr* audioMgr) {
     AudioTask* rspTask;
 
-    if (SREG(20) > 0) {
+    if (R_AUDIOMGR_DEBUG_LEVEL > AUDIOMGR_DEBUG_LEVEL_NONE) {
+        // Inhibit audio rsp task processing
         audioMgr->rspTask = NULL;
     }
+
     if (audioMgr->rspTask != NULL) {
+        // Got an rsp task to process, build the OSScTask and forward it to the scheduler to run
+
         audioMgr->audioTask.next = NULL;
         audioMgr->audioTask.flags = OS_SC_NEEDS_RSP;
         audioMgr->audioTask.framebuffer = NULL;
 
         audioMgr->audioTask.list = audioMgr->rspTask->task;
-        audioMgr->audioTask.msgQueue = &audioMgr->taskQueue;
+        audioMgr->audioTask.msgQueue = &audioMgr->taskDoneQueue;
 
         audioMgr->audioTask.msg = NULL;
         osSendMesg(&audioMgr->sched->cmdQueue, (OSMesg)&audioMgr->audioTask, OS_MESG_BLOCK);
         Sched_Notify(audioMgr->sched);
     }
 
+    // Update the audio driver
+
     gAudioThreadUpdateTimeStart = osGetTime();
-    if (SREG(20) >= 2) {
+
+    if (R_AUDIOMGR_DEBUG_LEVEL >= AUDIOMGR_DEBUG_LEVEL_NO_UPDATE) {
+        // Skip update, no rsp task produced
         rspTask = NULL;
     } else {
         rspTask = func_800E4FE0();
     }
+
     gAudioThreadUpdateTimeAcc += osGetTime() - gAudioThreadUpdateTimeStart;
     gAudioThreadUpdateTimeStart = 0;
 
     if (audioMgr->rspTask != NULL) {
-        osRecvMesg(&audioMgr->taskQueue, NULL, OS_MESG_BLOCK);
-        func_800C3C80(audioMgr);
+        // Wait for the audio rsp task scheduled on the previous retrace to complete. This looks like it should wait
+        // for the task scheduled on the current retrace, but since the queue is initially filled in AudioMgr_Init
+        // it doesn't wait for the task scheduler to post a message for the most recent task as there is already a
+        // message waiting.
+        osRecvMesg(&audioMgr->taskDoneQueue, NULL, OS_MESG_BLOCK);
+        // Report task done
+        //! @bug As a result of above, the task done notification is sent to the next task instead of the task it was
+        //! intended for. In practice, task done notifications are not used by the audio driver so this is
+        //! inconsequential.
+        AudioMgr_NotifyTaskDone(audioMgr);
```

Comment:
Is it? audioMgr->rspTask is still the previous task at this point isn't it?

Reply:
Here's what I think happens, in all the gory details

- The task done queue initially holds a message
- First retrace runs
  - `audioMgr->rspTask` is initially null, no task is scheduled
  - Audio driver produces an rsp task (task 1)
  - `audioMgr->rspTask` is still null, don't try to pull a message from the task done queue
  - `audioMgr->rspTask` is updated to the generated rsp task (task 1)
- Second retrace runs
  - `audioMgr->rspTask` is valid, the task is scheduled (task 1)
  - Audio driver produces an rsp task (task 2)
  - `audioMgr->rspTask` is valid, try to read a message. This always succeeds immediately regardless of whether the task actually completed due to the initially full queue. Further, since the queue is full the scheduler blocks waiting for the queue to empty, so the message that task 1 gets is guaranteed to be the "extra" message. The scheduler will then refill the queue with the task done message for task 1 once it actually completes.
  - `audioMgr->rspTask` is updated to the generated rsp task (task 2)
- Third retrace runs
  - `audioMgr->rspTask` is valid, the task is scheduled (task 2)
  - Audio driver produces an rsp task (task 3)
  - `audioMgr->rspTask` is valid, try to read a message. This will read the task done message for task 1 and post it to task 2 since they use the same message queue and the scheduler will have posted the task done message for task 1 before task 2 can start, as the scheduler blocks if the task done queue is full.
  - `audioMgr->rspTask` is updated to the generated rsp task (task 3)
- Repeat

I think my wording can definitely be improved here. If you see a good way to make this more clear let me know.

Reply:
ah I see I missed that the generated task is kept for scheduling on the next retrace instead of being scheduled immediately (why?)

I don't think your wording can be improved as much as the code in the first place 8)

maybe instead of

> As a result of above, the task done notification is sent to the next task instead of the task it was intended for

, ? :

> As a result of above, the initial task done notification is sent before the first task is actually done, and then every task done notification is early by one task.

I don't like "is early by one task" but idk how to put it concisely

Reply:
:thinking: 

> As a result of above, the first task done notification is sent before the first task is actually done, then the first task being done causes sending a second notification before the second task is actually done, and so on. So every task done notification is early by one task.

Reply:
> ah I see I missed that the generated task is kept for scheduling on the next retrace instead of being scheduled immediately (why?)

I guess it could make the timing more consistent, since this way the rsp task start time is not sensitive to variation in how long the audio driver takes to update?

I tried to adjust the comment, not sure if it's much better though